File: strategies/kubectl.go
// Copyright 2025 Naren Yellavula
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package strategies

// KubectlHelpStrategy handles kubectl commands with sub-commands
type KubectlHelpStrategy struct {
	cmdRunner *CommandRunner
}

func NewKubectlHelpStrategy(cmdRunner *CommandRunner) *KubectlHelpStrategy {
	return &KubectlHelpStrategy{cmdRunner: cmdRunner}
}

func (k *KubectlHelpStrategy) SupportsCommand(baseCmd string) bool {
	return baseCmd == "kubectl"
}

func (k *KubectlHelpStrategy) Priority() int {
	return 2
}

func (k *KubectlHelpStrategy) GetHelp(cmdParts []string) (string, error) {
	cmd := NewCommand(cmdParts)

	if !cmd.HasSubCommand(1) {
		return k.cmdRunner.Run("kubectl", "--help")
	}

	// Handle kubectl subcommand help - supports multiple levels
	args := append(cmd.SubCmds, "--help")
	return k.cmdRunner.Run("kubectl", args...)
}

File: recaller-rs/src/help_system/strategies/kubectl.rs
use std::sync::Arc;

use anyhow::Result;

use crate::help_system::runner::CommandRunner;
use crate::help_system::strategy::{CommandParts, HelpStrategy};

pub struct KubectlHelpStrategy {
    runner: Arc<CommandRunner>,
}

impl KubectlHelpStrategy {
    pub fn new(runner: Arc<CommandRunner>) -> Self {
        Self { runner }
    }
}

impl HelpStrategy for KubectlHelpStrategy {
    fn priority(&self) -> i32 {
        2
    }

    fn supports_command(&self, base_cmd: &str) -> bool {
        base_cmd == "kubectl"
    }

    fn get_help(&self, command: &CommandParts) -> Result<String> {
        if !command.has_sub_command(1) {
            return self.runner.run("kubectl", &["--help"]);
        }

        let mut args: Vec<&str> = command.sub_cmds().iter().map(|s| s.as_str()).collect();
        args.push("--help");
        self.runner.run("kubectl", &args)
    }
}
